core/src/main/java/com/google/instrumentation/trace/Samplers.java

```diff
@@ -83,4 +97,70 @@ public String toString() {
       return "NeverSampleSampler";
     }
   }
+
+  private static final class ProbabilitySampler extends Sampler {
+    // We assume the lower 64 bits of the traceId's are randomly distributed around the whole (long)
+    // range. We convert an incoming probability into an upper bound on that value, such that we can
+    // just compare the absolute value of the id and the bound to see if we are within the desired
+    // probability range.  Using the low bits of the traceId also ensures that systems that only use
+    // 64 bit ID's will also work with this sampler.
+    private final long idUpperBound;
+    private final double probability;
+
+    /**
+     * Create a new {@link ProbabilitySampler}. The probability of sampling a trace is equal to that
+     * of the specified probability.
+     *
+     * @param probability The desired probability of sampling. Must be within [0.0, 1.0].
+     * @throws IllegalArgumentException if {@code probability} is out of range
+     */
+    private ProbabilitySampler(double probability) {
+      this.probability = probability;
+      if (probability < 0.0 || probability > 1.0) {
+        throw new IllegalArgumentException("probability must be in range [0.0, 1.0]");
+      }
+      // Special case the limits, to avoid any possible issues with lack of precision across
+      // double/long boundaries. For probability == 0.0, we use Long.MIN_VALUE as this guarantees
+      // that we will never sample a trace, even in the case where the id == Long.MIN_VALUE, since
+      // Math.Abs(Long.MIN_VALUE) == Long.MIN_VALUE.
+      if (probability == 0.0) {
+        idUpperBound = Long.MIN_VALUE;
+      } else if (probability == 1.0) {
+        idUpperBound = Long.MAX_VALUE;
+      } else {
+        idUpperBound = (long) (probability * Long.MAX_VALUE);
+      }
+    }
+
+    @Override
+    protected boolean shouldSample(
+        @Nullable SpanContext parentContext,
+        boolean remoteParent,
+        TraceId traceId,
+        SpanId spanId,
+        String name,
+        @Nullable List<Span> parentLinks) {
+      checkNotNull(traceId, "traceId");
+      checkNotNull(name, "name");
+      // Always enable sampling if parent was sampled.
+      if (parentContext != null && parentContext.getTraceOptions().isSampled()) {
+        return true;
+      }
+      // TODO(bdrutu): Fix this to use TraceId#compareTo.
+      BigInteger traceIdNum = new BigInteger(traceId.getBytes());
```

Comment:
I think the TODO is about avoiding an extra copy of the traceId bytes. In order to do that you may add a package protected method in the TraceId.getLowerLong() { return new BigInteger(bytes).longValue(); }. This will avoid one extra copy of the bytes in getBytes().

If you do this we can remove the TODO.